### DISCUSSION TITLE
Replacing forward slash (/) in a field when renaming files

### DISCUSSION THREAD

**Discussion by ygabuev** [beetbox/beets#4020] [Q&A]
Hi!

I have a bunch of albums where artists are listed as "Artist A / Artist B". When doing `beet move`, I want the artists to be displayed in the filenames as "Artist A & Artist B". To do that, I put the following snippet in the config:
```yaml
replace:
    # these two lines differ from the default config
    '[\\]': _
    '[/]': '&'

    '^\.': _
    '[\x00-\x1f]': _
    '[<>:"\?\*\|]': _
    '\.$': _
    '\s+$': ''
    '^\s+': ''
    '^-': _
```

However, this doesn't work - it still changes the forward slash (/) to an underscore (_). Is it a bug or am I doing somthing wrong? What should I do to replace forward slashes correctly?

**Comment by sampsyo** [comment]
You’re running into this specific and challenging issue: https://github.com/beetbox/beets/issues/323

**Reply by ygabuev** [reply]
Hi again, it took me some time to come back to this issue.

I am not particularly sure yet how #323 is related to my issue, but I looked into the the code and below is my current understanding of how things work.

When we move (rename) items, here is the order of function calls, in dirty pseudocode with nesting and simplifications:
- `Item.move()`
  - `dest = Item.destination()`
    - `subpath = Item.evaluate_template(subpath_template, for_path=True)`
      - `return subpath_template.substitute(self.formatted(for_path=True), self._template_funcs())`
        - ...
    - `subpath = unicodedata.normalize(subpath)`
    - `subpath = util.asciify_path(subpath)`
    - `subpath = util.legalize_path(subpath, replacements, maxlen, ...)`
    - `return some_func(subpath)`

Currently, all of the substitutions are done during `util.legalize_path()` function call, where we already have a full-path string with path separators. It seems that this is one of the reasons why it is complicated to replace slashes as they are path separators on many platforms.

What if we did the replacements somewhere around `evaluate_template()`? We will first do the replacements on the per-field basis and then we can populate the template.

**Reply by sampsyo** [reply]
Yes, you've definitely identified the problem that #323 would need to address. Unfortunately, moving the legalization to `evaluate_template` won't work—replacement need to happen on fully-assembled path components because they might be sensitive to the context. For example, replacements that consist of multiple characters could span two field expansions (e.g., if you have `$foo$bar` where `foo`'s last character combined with `bar`'s first character creates an illegal sequence). Or, more realistically, some patterns are only legal or illegal at the beginning or end of a path component, such as in the pattern `^\.`, which wouldn't be possible to detect at that point.

It's a very tricky issue!

**Reply by ygabuev** [reply]
Oh I see. I looked into `functemplate.py` and that's indeed not that simple. So many edge cases, espicially due to the ability to call functions! Probably easier to use `path_sep_replase`.

However, there is still no info on it in the documentation. Would you like a small PR with minimal documentation in `reference/config.rst`?

**Reply by sampsyo** [reply]
I think that makes sense! One thing that would make sense for the docs is to include a strong warning… I am a tad worried that people could really screw things up on their filesystem by, for example, setting that option to `/`. 🤷

**Reply by ygabuev** [reply]
Oh, and also the function `util.asciify_path()` doesn't seem to actually use `sep_replace` argument, because we split the path into components on `os.sep`, so we are left with components that don't have separators in them. The same goes for `os.altsep`, because we transform it into `os.sep` as the first step. Am I correct? If yes, I can modify the code and probably add some tests.

**Reply by sampsyo** [reply]
That does sound right, at least (admittedly) without looking into it too deeply… yes, I'd be happy to read some draft code!

**Reply by ygabuev** [reply]
Turned out that I was wrong in my previous message, because the separators can appear after decoding the string. So, nothing to change.

Still, added some rudimentery documentation for `path_sep_replace` in https://github.com/beetbox/beets/pull/4119.

**Comment by 2V3EvG4LMJFdRe** [comment]
Similar to this, what do I put in `replace` to make beets use "..." instead of the single-character "…"? I also want this to account for problematic filenames (do "_.." in the filename), and to work on the tags of the file too (always do "...").
